Charachter controllers

AlexanderEvans opened the pull request:
changed how the music and enemy listing logic gets handled. Still bad, but better than it was. Will need a more complete refactoring sometime in the future yet. Still, it's a start.

Commit:
Attempting to fix repo's git configuration(.gitignore & .gitattributes)

Commit:
Enemy & player controller refactor

Dreaming381 commented on ComfySoda_Unity2018.2.7f1/Assets/Scripts/ComySodaEngine/EnemyController.cs:
```diff
@@ -0,0 +1,92 @@
+
+namespace cs
+{
+    public class EnemyController : MonoBehaviour
```
"Controller" is typically a per instance thing whereas "Manager" or "System" are usually singletons.

AlexanderEvans replied:
Unity refers to other things as "Manager" which is why I was trying to move away from that naming convention.  I might adopt "System", but then if I do hybrid ECS I could run into problems later.  I will overhaul all the class names in the project in a future refactor though to enforce consistency.

Dreaming381 replied:
That's fair. I recently had this discussion with Genzo too. We came up with a pretty funny name that he stuck with, but I forget what it was. "Supervisor" is a decent name for these types of things too.

As for ECS, recent developments have made it more of an All-or-Nothing approach, even with hybrid ECS. For your use case I think sticking with classical for the year makes the most sense. If you want to learn ECS, you can do that with your own private project, in which I will encourage you to try pure ECS because processing over a million entities in realtime is freaking fun!

AlexanderEvans replied:
If unity added the same level of support for things like physics in ECS I would consider it because I actually really like functional programming for such systems, but the lack of engine tool integration for things like polygonal physics and tilemap integration will have me hold off until the tools are more mature unfortunately.  Supervisor is good.  I will probably move to that in the next big refactor of this repo(the way things interconnect in it atm is a mess, I plan to refactor it to be more similar to magnum prandium in structure).  I'm still trying to decide on what variation of architecture I like, but I think what I ultimately go with will at least be similar to what I am currently attempting in magnum prandium.  There are just a few more things I want to straighten out with how I handle scene data that I don't really like at the moment.  It feels a bit silly to make a whole component for what seems to typically amount to a list and an int, but I really don't want to just make them statics, because they should vanish with the scene on scene changes, and if I am doing asynchronous loads/unloads that might become an issue for things that are meant to be contained within a scene, and scriptable objects aren't really any better of a solution for the same reason imo...  idk, I feel like I'm onto something with the current architecture in magnum prandium though.

Dreaming381 commented on ComfySoda_Unity2018.2.7f1/Assets/Scripts/ComySodaEngine/EnemyController.cs:
```diff
@@ -0,0 +1,92 @@
+        [SerializeField]
+        public float CLOSE = 7;
+        [SerializeField]
+        public float MID = 12;
```
public on any Serializable class (which Monobehaviours always are) is serializable by default. This is redundant.

AlexanderEvans replied:
yeah, that was left over code from testing I missed when it was private.  I've removed it due to redundancy as per your suggestion.

Dreaming381 commented on ComfySoda_Unity2018.2.7f1/Assets/Scripts/ComySodaEngine/EnemyController.cs:
```diff
@@ -0,0 +1,92 @@
+        public enum EnemyRange { CLOSE, MID, FAR};
+        public EnemyRange enemyRange = EnemyRange.FAR;
+        public List<EnemyLogic> enemyList = EnemyLogic.enemyList;
+
```
These also are serialized. Given your naming convention, I don't think you want that. Otherwise your naming convention is nonexistent.

AlexanderEvans replied:
serialization is undesirable here, but it needs to be referenced by other classes.

Dreaming381 replied:
[Nonserialized] ? Or make it an autoproperty (which Unity doesn't serialize). Autoproperties also let you make things public readable but private for writing, which in many cases is useful. Or you can decide to change the property behavior to trigger events on changes. Or even cooler yet, you can make them part of an interface.

Dreaming381 commented on ComfySoda_Unity2018.2.7f1/Assets/Scripts/ComySodaEngine/EnemyController.cs:
```diff
@@ -0,0 +1,92 @@
+                }
+            }
+        }
+        #endregion
```
I don't think this code works the way you expect it to. This code only applies closeness logic to the last enemy it checks. You'll need to move your EnemRange.Far initialization out of the loop. Also, what's up with the random scoping around the if/else block?

AlexanderEvans replied:
You're right, fixed.  and scoping... oh, LOL.  I think my tab spacing got off somehow when I was trying to figure out if I could do it with a nested loop and I just glanced through it and decided float tempDistance was a function in my head.  I'll remove those, they don't need to be there, haha.

Dreaming381 commented on ComfySoda_Unity2018.2.7f1/Assets/Scripts/ComySodaEngine/EnemyController.cs:
```diff
@@ -0,0 +1,92 @@
+                case EnemyRange.FAR:
+                    MusicManager.musicState = MusicState.main;
+                    break;
+            }
```
You have code duplication here. Move the game over logic outside the switch.

AlexanderEvans replied:
done

Dreaming381 replied:
You missed the `if` in `case EnemyRange.MID`.

Dreaming381 commented on ComfySoda_Unity2018.2.7f1/Assets/Scripts/ComySodaEngine/EnemyLogic.cs:
```diff
@@ -21,21 +21,33 @@ public class EnemyLogic : MonoBehaviour
+        [SerializeField]
         Waypoint nextTargetedWaypoint = null;
         Rigidbody2D myRigidbody2D;
         Animator tempAnimator;
```
Why is the Animator temporary?

AlexanderEvans replied:
because at one point it was temporary and I never renamed it.  It's fixed now.  Thanks.

Dreaming381 commented on ComfySoda_Unity2018.2.7f1/Assets/Scripts/ComySodaEngine/GameOverController.cs:
```diff
@@ -40,6 +40,7 @@ private void OnDestroy()
         public static void StartGameOver()
         {
             function();
```
This is the stupidest singleton I have ever seen. Declare a proper static instance. Or make private static events that your instance subscribes to and your static methods invoke. But this is just stupid. I know SDSU has a professor that teaches this in C++. It is bad practice in C++. It is worse in C#.

AlexanderEvans replied:
I agree it is stupid, I implemented it as a workaround.  I have it coded this way because I need to call StartCoroutine() which can only be called from an instance of a monobehaviour: https://imgur.com/a/ja3682z

Dreaming381 replied:
The following is a proper static instance implementation that will do what you want to do without bad hacks. I will leave it up to you to explore the private static events - instance subscriber approach if you choose to do so.
``` csharp
private static GameOverController gameOverController;
...
public static void StartGameOver()
{
    gameOverController.StartCoroutine(GameOverDisplay());
}
```

AlexanderEvans replied:
Oh yeah, that would work too, and it even looks cleaner, I **should** do it that way.  When I get around to refactoring this again I will definitely make this change.

Dreaming381 reviewed (commented):
I can only comment on the diff. It might be worth investing some time into a proper collection of game managers that do different things. Single Responsibility Principle is really important in classical Unity.

Commit:
fixed most comments in pull request

still need to refactor static singleton function()

Dreaming381 reviewed (commented):
Getting better. A couple of things you can still learn yet. But I think you'll be really happy with this once you get through the process a couple of times. I hope you find you are learning new things a lot faster. That's where code reviews are helpful is they point out hacks and show better ways to solve the problem using techniques you never knew were things. At least that's been my experience.

Commit:
Refactored names and adjusted dumb code.

Refactored names and adjusted dumb code.  Also extended dynamically adding characters to the game(serialization support for dictionaries and hash tables is missing so I implemented a list with a custom property drawer and load it up on awake).

Commit:
Abstracted cutscene sprites and dialogue replacemement strings

StoryCharachterGameSupervisor was implemented to handle all the game's dialogue sprites, and cutscene IDTag to DisplayName mappings.
